go-update/src/main/java/org/reactome/release/goupdate/GoLineProcessor.java

```diff
@@ -0,0 +1,232 @@
+/**
+ * 
+ */
+package org.reactome.release.goupdate;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * @author sshorser
+ *
+ */
+class GoLineProcessor
+{
+	private static final Logger logger = LogManager.getLogger();
+	
+	private GoLineProcessor() {}
+	
+	/**
+	 * Process a line from the GO file.
+	 * @param line - The line.
+	 * @param goTerms - The GO terms map. This map will be updated by this function.
+	 * @param currentGOID - The ID of the GO term currently being processed, line by line.
+	 * @return The GO Accession of the GO term currently being processed. Will be different from <code>currentGOID</code> if the ID for a new GO term is seen on this <code>line</code>
+	 */
+	static String processLine(String line, String currentGOID, Map<String, Map<String,Object>> goTerms)
+	{
+		String goID = currentGOID;
+		Matcher m;
+		try
+		{
+			m = GoUpdateConstants.LINE_DECODER.matcher(line);
+			if (m.matches())
+			{
+				String lineCode = m.group(1);
+				switch (lineCode)
+				{
+					case GoUpdateConstants.ID:
+					{
+						m = GoUpdateConstants.GO_ID_REGEX.matcher(line);
+						goID = m.matches() ? m.group(1) : "";
+						// Were we able to extract a GO ID?
+						if (!goID.trim().isEmpty())
+						{
+							if (!goTerms.containsKey(goID))
+							{
+								goTerms.put(goID, new HashMap<String,Object>());
+								currentGOID = goID;
+							}
+							else
+							{
+								// If a GO ID appears a second time, it will not be added to the hash, and the user will 
+								// get a message asking them to verify if the file is really OK.
+								// Maybe throw a RuntimeException for this? It really should never happen.
+								logger.error("GO ID {} has appeared more than once in the input! This is highly unexpected. "
```

Comment:
Is it rare enough and bad enough that it warrants the runtime exception? Or could it be flagged and filtered out

Reply:
It never actually happened in my testing with real data. I don't think the old code handled it all. Currently, the code will just log an error message. That code was just a thought about throwing a `RuntimeException` if it happens. But maybe I should actually add that in.
The reason? If a GO Id actually did appear twice in the file, and the content for each occurrence was different, how would you decide which one to keep and which to discard?

Reply:
Yea I suppose in that case you would want to make a note of it and potentially remove both. I guess we'll cross that bridge when it appears?

Reply:
Right now, it would actually keep the first one it finds and all others would trigger an error message. That seems arbitrary, so I think I *will* add that RuntimeException.